### ISSUE TITLE
Unable to disable warning when using derive

### ISSUE THREAD

**Issue by alexkazik** [rust-lang/rust-clippy#10556]
### Summary

I want to silence a clippy warning.

I did stick the allow to all places I thought possible.

It is silenced without the derive.

### Lint Name

type_complexity

### Reproducer

I tried this code:

```rust
use std::borrow::Cow;

#[derive(serde::Serialize, serde::Deserialize)]
pub enum EncounterType {}

#[allow(clippy::type_complexity)]
#[derive(serde::Serialize, serde::Deserialize)]
#[allow(clippy::type_complexity)]
pub struct SerdeEncounter<'a>(
    #[allow(clippy::type_complexity)] Vec<(Cow<'a, str>, Vec<EncounterType>, usize, Cow<'a, str>)>,
);
```

I saw this happen:

```
warning: very complex type used. Consider factoring parts into `type` definitions
  --> app/src/lib.rs:99:39
   |
99 |     #[allow(clippy::type_complexity)] Vec<(Cow<'a, str>, Vec<EncounterType>, usize, Cow<'a, str>)>,
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#type_complexity
```

I expected to see this happen:
Nothing

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=779725236b9bfeccf85248c64f8acd49

### Version

```text
tested on play.rust-lang.org, which reports:
stable rust 1.68.1
clippy 0.1.70 (2023-03-26 db0cbc4)
```

**Comment by bjorn3** [comment]
The serde::Serialize` and `serde::Deserialize` proc macros expand into trait implementations that mention this complex type without applying the `#[allow(clippy::type_complexity)]` attribute on the trait implementation. Rustc warns about the usage inside the derived implementation, not about the field directly. Looks like it does, because the proc macros use the span of the field type as the span for each mention of this type inside the derived implementation. For reference this is what the expansion looks like:

```rust
#[allow(clippy::type_complexity)]
#[allow(clippy::type_complexity)]
pub struct SerdeEncounter<'a>(
    #[allow(clippy::type_complexity)] Vec<(Cow<'a, str>, Vec<EncounterType>, usize, Cow<'a, str>)>,
);
#[doc(hidden)]
#[allow(non_upper_case_globals, unused_attributes, unused_qualifications)]
const _: () = {
    #[allow(unused_extern_crates, clippy::useless_attribute)]
    extern crate serde as _serde;
    #[automatically_derived]
    impl<'a> _serde::Serialize for SerdeEncounter<'a> {
        fn serialize<__S>(
            &self,
            __serializer: __S,
        ) -> _serde::__private::Result<__S::Ok, __S::Error>
        where
            __S: _serde::Serializer,
        {
            _serde::Serializer::serialize_newtype_struct(__serializer, "SerdeEncounter", &self.0)
        }
    }
};
#[doc(hidden)]
#[allow(non_upper_case_globals, unused_attributes, unused_qualifications)]
const _: () = {
    #[allow(unused_extern_crates, clippy::useless_attribute)]
    extern crate serde as _serde;
    #[automatically_derived]
    impl<'de, 'a> _serde::Deserialize<'de> for SerdeEncounter<'a> {
        fn deserialize<__D>(__deserializer: __D) -> _serde::__private::Result<Self, __D::Error>
        where
            __D: _serde::Deserializer<'de>,
        {
            #[doc(hidden)]
            struct __Visitor<'de, 'a> {
                marker: _serde::__private::PhantomData<SerdeEncounter<'a>>,
                lifetime: _serde::__private::PhantomData<&'de ()>,
            }
            impl<'de, 'a> _serde::de::Visitor<'de> for __Visitor<'de, 'a> {
                type Value = SerdeEncounter<'a>;
                fn expecting(
                    &self,
                    __formatter: &mut _serde::__private::Formatter,
                ) -> _serde::__private::fmt::Result {
                    _serde::__private::Formatter::write_str(
                        __formatter,
                        "tuple struct SerdeEncounter",
                    )
                }
                #[inline]
                fn visit_newtype_struct<__E>(
                    self,
                    __e: __E,
                ) -> _serde::__private::Result<Self::Value, __E::Error>
                where
                    __E: _serde::Deserializer<'de>,
                {
                    let __field0:
                                Vec<(Cow<'a, str>, Vec<EncounterType>, usize,
                                Cow<'a, str>)> =
                            match <Vec<(Cow<'a, str>, Vec<EncounterType>, usize,
                                        Cow<'a, str>)> as _serde::Deserialize>::deserialize(__e) {
                                _serde::__private::Ok(__val) => __val,
                                _serde::__private::Err(__err) => {
                                    return _serde::__private::Err(__err);
                                }
                            };
                    _serde::__private::Ok(SerdeEncounter(__field0))
                }
                #[inline]
                fn visit_seq<__A>(
                    self,
                    mut __seq: __A,
                ) -> _serde::__private::Result<Self::Value, __A::Error>
                where
                    __A: _serde::de::SeqAccess<'de>,
                {
                    let __field0 = match match _serde::de::SeqAccess::next_element::<
                        Vec<(Cow<'a, str>, Vec<EncounterType>, usize, Cow<'a, str>)>,
                    >(&mut __seq)
                    {
                        _serde::__private::Ok(__val) => __val,
                        _serde::__private::Err(__err) => {
                            return _serde::__private::Err(__err);
                        }
                    } {
                        _serde::__private::Some(__value) => __value,
                        _serde::__private::None => {
                            return _serde::__private::Err(_serde::de::Error::invalid_length(
                                0usize,
                                &"tuple struct SerdeEncounter with 1 element",
                            ));
                        }
                    };
                    _serde::__private::Ok(SerdeEncounter(__field0))
                }
            }
            _serde::Deserializer::deserialize_newtype_struct(
                __deserializer,
                "SerdeEncounter",
                __Visitor {
                    marker: _serde::__private::PhantomData::<SerdeEncounter<'a>>,
                    lifetime: _serde::__private::PhantomData,
                },
            )
        }
    }
};
```

If you run clippy on it, you will see that the warning is about the `let __field0:
                                Vec<(Cow<'a, str>, Vec<EncounterType>, usize,
                                Cow<'a, str>)> =` part.

**Comment by alexkazik** [comment]
Wow, I did looked at cargo expand just to make sure the attributes are still there but didn't try to run it. Thanks.

**Comment by Jarcho** [comment]
@blyxyas Just a heads up. You'll need to add a check if the lint type is found within an `automatically_derived` impl. `is_from_proc_macro` won't work here.
